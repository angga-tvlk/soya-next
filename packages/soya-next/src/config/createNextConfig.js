export default ({ assetPrefix = '', webpack, ...config } = {}) => ({
  ...config,
  assetPrefix,
  webpack(webpackConfig, { dev }) {
    let localIdentName, imageName;
    if (dev) {
      localIdentName = '[name]__[local]--[hash:base64:5]';
      imageName = '[name]';
    } else {
      localIdentName = null;
      imageName = '[name]-[hash]';
    }

    const cssLoaderOptions = {
      localIdentName,
      minimize: !dev,
      sourceMap: dev,
    };

    const rule = webpackConfig.module.rules.find(rule => (
      rule.loader === 'babel-loader' &&
      rule.test && rule.test.toString() === '/\\.js(\\?[^?]*)?$/'
    ));
    if (rule && !rule.options.babelrc) {
      // istanbul ignore else
      if (process.env.NODE_ENV === 'test') {
        rule.options.presets.push('../babel/preset');
      } else {
        rule.options.presets.push(require.resolve('../babel/preset'));
      }
    }
    webpackConfig.module.rules.push(
      {
        test: [
          /\.(css|s(a|c)ss)$/,
          /\.(bmp|gif|jpe?g|png)$/,
        ],
        loader: 'emit-file-loader',
        options: {
          name: 'dist/[path][name].[ext]',
        },
      },
      {
        test: /\.mod(ule)?\.css$/,
        use: [
          'babel-loader',
          'styled-modules/loader',
          {
            loader: 'css-loader',
            options: {
              ...cssLoaderOptions,
              modules: true,
            },
          },
        ],
      },
      {
        test: /\.mod(ule)?\.s(a|c)ss$/,
        use: [
          'babel-loader',
          'styled-modules/loader',
          {
            loader: 'css-loader',
            options: {
              ...cssLoaderOptions,
              modules: true,
              importLoaders: 1,
            },
          },
          'sass-loader',
        ],
      },
      {
        test: /\.css$/,
        exclude: /\.mod(ule)?\.css$/,
        use: [
          'babel-loader',
          'styled-modules/loader',
          {
            loader: 'css-loader',
            options: cssLoaderOptions,
          },
        ],
      },
      {
        test: /\.s(a|c)ss$/,
        exclude: /\.mod(ule)?\.s(a|c)ss$/,
        use: [
          'babel-loader',
          'styled-modules/loader',
          {
            loader: 'css-loader',
            options: {
              ...cssLoaderOptions,
              importLoaders: 1,
            },
          },
          'sass-loader',
        ],
      },
      {
        test: /\.(bmp|gif|jpe?g|png|ttf|eot|woff2?|svg)$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: `dist/static/[path]${imageName}.[ext]`,
          publicPath: url => `${assetPrefix}/_soya/${url.replace('dist/static/', '')}`,
        },
      },
    );

    if (webpack) {
      webpack(webpackConfig, { dev });
    }

    return webpackConfig;
  },
});
