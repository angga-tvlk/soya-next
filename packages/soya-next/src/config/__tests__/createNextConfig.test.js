import createNextConfig from '../createNextConfig';

let webpackConfig;

describe('Create Next Config', () => {
  it('should return default config', () => {
    expect(createNextConfig()).toMatchSnapshot();
  });

  describe('Webpack Config', () => {
    beforeEach(() => {
      webpackConfig = {
        module: {
          rules: [{
            test: /\.js(\?[^?]*)?$/,
            loader: 'babel-loader',
            options: {
              presets: [],
              babelrc: false,
            },
          }],
        },
        plugins: [],
      };
    });

    it('should return default development config', () => {
      const { webpack } = createNextConfig();
      expect(webpack(webpackConfig, { dev: true })).toMatchSnapshot();
    });

    it('should return default production config', () => {
      const { webpack } = createNextConfig();
      expect(webpack(webpackConfig, { dev: false })).toMatchSnapshot();
    });

    it('should not add any presets if babelrc is used', () => {
      const { webpack } = createNextConfig();
      const babelLoaderRule = webpackConfig.module.rules.find(({ loader }) => loader === 'babel-loader');
      babelLoaderRule.options.babelrc = true;
      expect(webpack(webpackConfig, { dev: false })).toMatchSnapshot();
    });

    it('should return customized config', () => {
      const { webpack } = createNextConfig({
        webpack(config) {
          config.plugins.push(jest.fn());
        },
      });
      expect(webpack(webpackConfig, { dev: false })).toMatchSnapshot();
    });

    it('should return correct public path url', () => {
      const { webpack } = createNextConfig({
        webpack(config) {
          config.plugins.push(jest.fn());
        },
      });
      const urlLoaderRule = webpack(webpackConfig, { dev: false }).module.rules.find(({ loader }) => loader === 'url-loader');
      expect(urlLoaderRule.options.publicPath('dist/static/path/to/soya.png')).toBe('/_soya/path/to/soya.png');
    });
  });
});
