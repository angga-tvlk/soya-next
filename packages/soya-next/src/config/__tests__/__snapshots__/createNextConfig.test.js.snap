// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Next Config Webpack Config should not add any presets if babelrc is used 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "loader": "babel-loader",
        "options": Object {
          "babelrc": true,
          "presets": Array [],
        },
        "test": /\\\\\\.js\\(\\\\\\?\\[\\^\\?\\]\\*\\)\\?\\$/,
      },
      Object {
        "loader": "emit-file-loader",
        "options": Object {
          "name": "dist/[path][name].[ext]",
        },
        "test": Array [
          /\\\\\\.\\(css\\|s\\(a\\|c\\)ss\\)\\$/,
          /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\)\\$/,
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "loader": "url-loader",
        "options": Object {
          "limit": 10000,
          "name": "dist/static/[path][name]-[hash].[ext]",
          "publicPath": [Function],
        },
        "test": /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\|ttf\\|eot\\|woff2\\?\\|svg\\)\\$/,
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`Create Next Config Webpack Config should return customized config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "loader": "babel-loader",
        "options": Object {
          "babelrc": false,
          "presets": Array [
            "../babel/preset",
          ],
        },
        "test": /\\\\\\.js\\(\\\\\\?\\[\\^\\?\\]\\*\\)\\?\\$/,
      },
      Object {
        "loader": "emit-file-loader",
        "options": Object {
          "name": "dist/[path][name].[ext]",
        },
        "test": Array [
          /\\\\\\.\\(css\\|s\\(a\\|c\\)ss\\)\\$/,
          /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\)\\$/,
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "loader": "url-loader",
        "options": Object {
          "limit": 10000,
          "name": "dist/static/[path][name]-[hash].[ext]",
          "publicPath": [Function],
        },
        "test": /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\|ttf\\|eot\\|woff2\\?\\|svg\\)\\$/,
      },
    ],
  },
  "plugins": Array [
    [Function],
  ],
}
`;

exports[`Create Next Config Webpack Config should return default development config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "loader": "babel-loader",
        "options": Object {
          "babelrc": false,
          "presets": Array [
            "../babel/preset",
          ],
        },
        "test": /\\\\\\.js\\(\\\\\\?\\[\\^\\?\\]\\*\\)\\?\\$/,
      },
      Object {
        "loader": "emit-file-loader",
        "options": Object {
          "name": "dist/[path][name].[ext]",
        },
        "test": Array [
          /\\\\\\.\\(css\\|s\\(a\\|c\\)ss\\)\\$/,
          /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\)\\$/,
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": "[name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": true,
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": true,
              "sourceMap": true,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": "[name]__[local]--[hash:base64:5]",
              "minimize": false,
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[name]__[local]--[hash:base64:5]",
              "minimize": false,
              "sourceMap": true,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "loader": "url-loader",
        "options": Object {
          "limit": 10000,
          "name": "dist/static/[path][name].[ext]",
          "publicPath": [Function],
        },
        "test": /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\|ttf\\|eot\\|woff2\\?\\|svg\\)\\$/,
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`Create Next Config Webpack Config should return default production config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "loader": "babel-loader",
        "options": Object {
          "babelrc": false,
          "presets": Array [
            "../babel/preset",
          ],
        },
        "test": /\\\\\\.js\\(\\\\\\?\\[\\^\\?\\]\\*\\)\\?\\$/,
      },
      Object {
        "loader": "emit-file-loader",
        "options": Object {
          "name": "dist/[path][name].[ext]",
        },
        "test": Array [
          /\\\\\\.\\(css\\|s\\(a\\|c\\)ss\\)\\$/,
          /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\)\\$/,
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.css\\$/,
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "exclude": /\\\\\\.mod\\(ule\\)\\?\\\\\\.s\\(a\\|c\\)ss\\$/,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": Array [
          "babel-loader",
          "styled-modules/loader",
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": null,
              "minimize": true,
              "sourceMap": false,
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "loader": "url-loader",
        "options": Object {
          "limit": 10000,
          "name": "dist/static/[path][name]-[hash].[ext]",
          "publicPath": [Function],
        },
        "test": /\\\\\\.\\(bmp\\|gif\\|jpe\\?g\\|png\\|ttf\\|eot\\|woff2\\?\\|svg\\)\\$/,
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`Create Next Config should return default config 1`] = `
Object {
  "assetPrefix": "",
  "webpack": [Function],
}
`;
